src/readers/lsb4.c

```diff
@@ -0,0 +1,155 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#include <readers.h>
+#include <logs.h>
+
+#define EXTENSION_BLOCK_LENGTH 8
+
+static uint32_t get_length(FILE *file);
+
+Stego *retrieve_lsb4(FILE *file, size_t offset, char **extension)
+{
+    if (fseek(file, offset, SEEK_SET) < 0)
+    {
+        perror("File too small for offset");
+        fclose(file);
+        return NULL;
+    }
+
+    const uint32_t message_length = get_length(file);
+    if (message_length == 0)
+    {
+        fclose(file);
+        return NULL;
+    }
+
+    LOG("Message length: %u\n", message_length);
+
+    // Allocate memory for the message
+    uint8_t *message = malloc(message_length);
+    if (message == NULL)
+    {
+        perror("Memory allocation for message failed");
+        fclose(file);
+        return NULL;
+    }
+
+    // Read the message bits
+    for (size_t i = 0; i < message_length * 2; i++)
+    {
+        char byte = fgetc(file);
+
+        if (feof(file))
+        {
+            perror("Not enough bytes to read message");
+            LOG("Reached EOF at byte %zu\n", i);
+
+            free(message);
+            fclose(file);
+
+            return NULL;
+        }
+
+        message[i / 2] <<= 4;
+        message[i / 2] |= (byte & 0x0F);
+    }
+
+    LOG("Message: %.*s\n", message_length, message);
+
+    Stego *stego = malloc(sizeof(Stego) + message_length);
+    if (stego == NULL)
+    {
+        perror("Memory allocation for Stego failed");
+
+        free(message);
+        fclose(file);
+
+        return NULL;
+    }
+
+    stego->size = message_length;
+    memcpy(stego->data, message, message_length);
```

Comment:
We could be allocating the stego data at the begining of the code and store the values directly on it, instead of using temp variables which are almost immediately freed.